provider/pdns.go

```diff
@@ -0,0 +1,369 @@
+
+	pgo "github.com/ffledgling/pdns-go"
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+	"github.com/kubernetes-incubator/external-dns/plan"
+)
+
+type pdnsChangeType string
+
+const (
+	apiBase = "/api/v1"
+
+	// Unless we use something like pdnsproxy (discontinued upsteam), this value will _always_ be localhost
+	defaultServerID = "localhost"
+	defaultTTL      = 300
+
+	// PdnsDelete and PdnsReplace are effectively an enum for "pgo.RrSet.changetype"
+	// TODO: Can we somehow get this from the pgo swagger client library itself?
+
+	// PdnsDelete : PowerDNS changetype used for deleting rrsets
+	// ref: https://doc.powerdns.com/authoritative/http-api/zone.html#rrset (see "changetype")
+	PdnsDelete pdnsChangeType = "DELETE"
+	// PdnsReplace : PowerDNS changetype for creating, updating and patching rrsets
+	PdnsReplace pdnsChangeType = "REPLACE"
+)
+
+// Function for debug printing
+func stringifyHTTPResponseBody(r *http.Response) (body string) {
+
+	if r == nil {
+		return ""
+	}
+
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(r.Body)
+	body = buf.String()
+	return body
+
+}
+
+// PDNSAPIProvider : Interface used and extended by the PDNSAPIClient struct as
+// well as mock APIClients used in testing
+type PDNSAPIProvider interface {
+	ListZones() ([]pgo.Zone, *http.Response, error)
+	ListZone(zoneID string) (pgo.Zone, *http.Response, error)
+	PatchZone(zoneID string, zoneStruct pgo.Zone) (*http.Response, error)
+}
+
+// PDNSAPIClient : Struct that encapsulates all the PowerDNS specific implementation details
+type PDNSAPIClient struct {
+	dryRun  bool
+	authCtx context.Context
+	client  *pgo.APIClient
+}
+
+// ListZones : Method returns all enabled zones from PowerDNS
+// ref: https://doc.powerdns.com/authoritative/http-api/zone.html#get--servers-server_id-zones
+func (c *PDNSAPIClient) ListZones() ([]pgo.Zone, *http.Response, error) {
+	zones, resp, err := c.client.ZonesApi.ListZones(c.authCtx, defaultServerID)
+	if err != nil {
+		log.Warnf("Unable to fetch zones. %v", err)
+	}
+	return zones, resp, err
+}
+
+// ListZone : Method returns the details of a specific zone from PowerDNS
+// ref: https://doc.powerdns.com/authoritative/http-api/zone.html#get--servers-server_id-zones-zone_id
+func (c *PDNSAPIClient) ListZone(zoneID string) (pgo.Zone, *http.Response, error) {
+	zones, resp, err := c.client.ZonesApi.ListZone(c.authCtx, defaultServerID, zoneID)
+	if err != nil {
+		log.Warnf("Unable to list zone %s. %v", zoneID, err)
+	}
+	return zones, resp, err
+}
+
+// PatchZone : Method used to update the contents of a particular zone from PowerDNS
+// ref: https://doc.powerdns.com/authoritative/http-api/zone.html#patch--servers-server_id-zones-zone_id
+func (c *PDNSAPIClient) PatchZone(zoneID string, zoneStruct pgo.Zone) (*http.Response, error) {
+	resp, err := c.client.ZonesApi.PatchZone(c.authCtx, defaultServerID, zoneID, zoneStruct)
+	if err != nil {
+		log.Warnf("Unable to patch zone %s. %v", zoneStruct.Name, err)
```

Comment:
Is it okay to only log this error?

Reply:
@Raffo Would you prefer this be a hard fail instead? Or are you suggesting it should be logged at level `info` ?

Reply:
I think it makes sense to return an error here, in the end we want to know that something went wrong. Also, I see that when we call `p.mutateRecords` we don't actually have any check for the error... is that intended and do I look at it right? If it is not intended, maybe a run of a linter would help to spot those cases.

Reply:
The reason I left it as a warning because network blips might result in one-off failures of the request, so it might make sense to retry instead of hard failing the first time we see an error. An alternative of-course would be to add retry logic w/ back-off so that we hard fail if we see no response after say 3 tries.

Reply:
please return errors and because api-sdks like aws do retries, you might want to do the retry inside the pdns provider.

Reply:
This now retries thrice and returns an error otherwise.